File: components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabView.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.rocket.tabs;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Message;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.View;

import org.mozilla.rocket.tabs.web.DownloadCallback;

public interface TabView {
    class HitTarget {
        public final TabView source;
        public final boolean isLink;
        public final String linkURL;

        public final boolean isImage;
        public final String imageURL;

        public HitTarget(@NonNull final TabView source,
                         final boolean isLink,
                         final String linkURL,
                         final boolean isImage,
                         final String imageURL) {

            if (isLink && linkURL == null) {
                throw new IllegalStateException("link hittarget must contain URL");
            }

            if (isImage && imageURL == null) {
                throw new IllegalStateException("image hittarget must contain URL");
            }

            this.source = source;
            this.isLink = isLink;
            this.linkURL = linkURL;
            this.isImage = isImage;
            this.imageURL = imageURL;
        }
    }

    interface FullscreenCallback {
        void fullScreenExited();
    }

    interface FindListener {
        void onFindResultReceived(int activeMatchOrdinal, int numberOfMatches, boolean isDoneCounting);
    }

    /**
     * Enable/Disable content blocking for this session (Only the blockers that are enabled in the app's settings will be turned on/off).
     */
    void setContentBlockingEnabled(boolean enabled);

    /**
     * Be invoked by TabChromeClient.onCreateWindow to transport this new-created-window to its parent window.
     *
     * @param msg The message to send when once a new WebView has been created
     */
    void bindOnNewWindowCreation(@NonNull final Message msg);

    void setImageBlockingEnabled(boolean enabled);

    boolean isBlockingEnabled();

    void performExitFullScreen();

    void setViewClient(@Nullable TabViewClient viewClient);

    void setChromeClient(@Nullable TabChromeClient chromeClient);

    void setDownloadCallback(DownloadCallback callback);

    void setFindListener(FindListener callback);

    void onPause();

    void onResume();

    void destroy();

    void reload();

    void stopLoading();

    String getUrl();

    String getTitle();

    @SiteIdentity.SecurityState
    int getSecurityState();

    void loadUrl(String url);

    void cleanup();

    void goForward();

    void goBack();

    boolean canGoForward();

    boolean canGoBack();

    void restoreViewState(Bundle inState);

    void saveViewState(Bundle outState);

    void insertBrowsingHistory();

    View getView();

    void buildDrawingCache(boolean autoScale);

    Bitmap getDrawingCache(boolean autoScale);
}

File: components/feature/tabs/src/main/java/org/mozilla/rocket/tabs/TabView.kt
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package org.mozilla.rocket.tabs

import android.graphics.Bitmap
import android.os.Bundle
import android.os.Message
import android.view.View
import org.mozilla.rocket.tabs.SiteIdentity.SecurityState
import org.mozilla.rocket.tabs.web.DownloadCallback

interface TabView {
    val view: View
    val isBlockingEnabled: Boolean
    val url: String?
    val title: String?

    @get:SecurityState
    val securityState: Int

    /**
     * Enable/Disable content blocking for this session (Only the blockers that are enabled in the app's settings will be turned on/off).
     */
    fun setContentBlockingEnabled(enabled: Boolean)

    /**
     * Be invoked by TabChromeClient.onCreateWindow to transport this new-created-window to its parent window.
     *
     * @param msg The message to send when once a new WebView has been created
     */
    fun bindOnNewWindowCreation(msg: Message)

    fun setImageBlockingEnabled(enabled: Boolean)
    fun setJavaScriptBlockingEnabled(enabled: Boolean)
    fun performExitFullScreen()
    fun setViewClient(viewClient: TabViewClient?)
    fun setChromeClient(chromeClient: TabChromeClient?)
    fun setDownloadCallback(callback: DownloadCallback?)
    fun setFindListener(callback: FindListener?)
    fun onPause()
    fun onResume()
    fun destroy()
    fun reload()
    fun stopLoading()
    fun loadUrl(url: String?)
    fun cleanup()
    fun goForward()
    fun goBack()
    fun canGoForward(): Boolean
    fun canGoBack(): Boolean
    fun restoreViewState(inState: Bundle?)
    fun saveViewState(outState: Bundle?)
    fun insertBrowsingHistory()
    fun buildDrawingCache(autoScale: Boolean)
    fun getDrawingCache(autoScale: Boolean): Bitmap?

    class HitTarget(
        source: TabView,
        isLink: Boolean,
        linkURL: String?,
        isImage: Boolean,
        imageURL: String?
    ) {
        @JvmField
        val source: TabView

        @JvmField
        val isLink: Boolean

        @JvmField
        val linkURL: String?

        @JvmField
        val isImage: Boolean

        @JvmField
        val imageURL: String?

        init {
            check(!(isLink && linkURL == null)) { "link hittarget must contain URL" }
            check(!(isImage && imageURL == null)) { "image hittarget must contain URL" }
            this.source = source
            this.isLink = isLink
            this.linkURL = linkURL
            this.isImage = isImage
            this.imageURL = imageURL
        }
    }

    interface FullscreenCallback {
        fun fullScreenExited()
    }

    interface FindListener {
        fun onFindResultReceived(
            activeMatchOrdinal: Int,
            numberOfMatches: Int,
            isDoneCounting: Boolean
        )
    }
}
